DOCSP-26368: FAQ entry for state should be open exception (#416)
mongodb/docs-java: source/faq.txt

closes all idle sockets and closes all sockets that are in
use as they are returned to the pool.

> Why does the application throw "IllegalStateException: state should be: open"?

Reviewer: Suggestion: make the question more oriented around the user

```suggestion
Why did I receive an "IllegalStateException: state should be: open" error?
```

Author: Thanks, I reworded since "receive" doesn't seem like a common action to describe exceptions.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> You may encounter this exception if you call an operation on a ``MongoClient``
> instance that closed its connections to MongoDB. The connections could be

Reviewer: Question: Does the client close the connections or does the application close the connections? It could be helpful to clarify the wording here. 

```suggestion
You may encounter this exception if you call an operation on a ``MongoClient``
instance with a closed connection to MongoDB. The connections could be
```

Author: I'm not sure I understand the question since they both are responsible for the closing of the connection depending on perspective. An application must call the code to instruct the client to close. The client must close the connection when that method is called. Could you clarify what you meant?

Reviewer: I think I meant to ask whether there were times when a client closes a connection on its own, without the application containing/calling the close method. In any case, i think you can keep the wording as is

Author: From what I understand, the client, which I'm assuming is MongoClient, can open and close connections based on load and network state (see [How Does Connection Pooling Work in the Java Driver](https://www.mongodb.com/docs/drivers/java/sync/current/faq/#how-does-connection-pooling-work-in-the-java-driver-)), but when you call an operation, whether a direct one such as ListDatabaseNames() or indirectly such as by calling find() on a MongoCollection, the MongoClient queues it up for the next available connection.

However, if you call close() on the instance, or perhaps another process signals for it to close its connections, it transitions into a state that is no longer "open" and does not accept any further operations.

> instance that closed its connections to MongoDB. The connections could be
> closed due to one of the following conditions:

Reviewer: ```suggestion
instance that closed its connections to MongoDB. The connections could be
closed for one of the following reasons:
```

> - The ``close()`` method was called on that ``MongoClient`` instance

Reviewer: ```suggestion
- You previously called the ``close()`` method on the ``MongoClient`` instance.
```

> - The ``MongoClient`` instance is out of scope

Reviewer: ```suggestion
- The ``MongoClient`` instance is out of scope.
```

> - The framework your application uses to manage your MongoDB operations, such
>   as Spring Boot, has closed the connection automatically.

Reviewer: ```suggestion
- The framework your application uses to manage your MongoDB operations, such
  as Spring Boot, automatically closed the connection.
```

Also suggestion: link to something related to Spring Boot

closes all idle sockets and closes all sockets that are in
use as they are returned to the pool.

How do I prevent the "IllegalStateException: state should be: open" error?
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You may encounter this exception if you call an operation on a ``MongoClient``
instance that closed its connections to MongoDB. The client's connections
can be closed for one of the following reasons:

- The ``close()`` method was called on that ``MongoClient`` instance prior
  to the operation call.

> - Another thread or process closed the connection. This can happen when

Reviewer 2: I don't understand what the difference is between the first and second bullet.  The only way to "close the connection" is to call the close method on the MongoClient.  The difference only seems to be where the close is happening (how close/far it is from the code that is getting the exception, whether in the same thread or a different thread.  

Also, it's not possible for another process to directly close a MongoClient.  It would have to happen indirectly through communication through a socket, or some signal, etc.

Also, it may be the

Author: I see. I assumed that [this response](https://www.mongodb.com/community/forums/t/illegalstateexception-state-should-be-open/223912/4) was referring to an executor service operating on a pool of threads.  Would you prefer I remove this list item to avoid any confusion? Is there any info you'd like to add? Thanks!

Reviewer 2: Yes, please remove.  You could point out that often MongoClient instances are singletons that are accessible by multiple threads, and so the code which closes the MongoClient may be distant from the code receiving the exception, perhaps in another thread of execution.

.. tip::

   The code that closes the ``MongoClient`` instance may be difficult to

>    locate in certain cases. The ``MongoClient`` uses the singleton pattern,

Reviewer 2: MongoClient doesn't use the singleton pattern itself.  It's just that frameworks are often configured to manage these instances as singletons.
